File: apps/api/src/vigia_api/api/v1/incidents.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, ConfigDict, Field

from ...domain.incidents import IncidentStatus, parse_detection_event
from ...container import incident_repository
from ...services.incidents import incident_to_dict
from ...security.csrf import require_csrf
from ...security.dependencies import get_current_user, require_permission
from ...security.origin import validate_origin_or_referer

router = APIRouter(tags=["incidents"])

DEFAULT_LIMIT = 50
MAX_LIMIT = 100


def _repository(request: Request | None):
    container = getattr(getattr(getattr(request, "app", None), "state", None), "container", None)
    return container.incident_repository if container is not None else incident_repository


def _paginate(items: list[Any], limit: int = DEFAULT_LIMIT, offset: int = 0) -> dict:
    limit = max(1, min(limit, MAX_LIMIT))
    offset = max(0, offset)
    total = len(items)
    sliced = items[offset: offset + limit]
    return {"items": sliced, "page_info": {"limit": limit, "offset": offset, "total": total, "has_next": offset + limit < total}}


class DetectionEventIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    organization_id: str
    event_id: str | None = None
    site_id: str | None = None
    camera_id: str
    zone_id: str
    worker_id: str | None = None
    event_type: str = "detection"
    severity: str
    timestamp: str | None = None
    confidence: float | None = None
    model_version: str | None = None
    evidence: dict | None = None
    summary: str | None = None
    metadata: dict = Field(default_factory=dict)


class IncidentReasonIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=1)


def _status_response(incident):
    return incident_to_dict(incident)


@router.post("/organizations/{organization_id}/incidents:ingest", status_code=201, dependencies=[Depends(require_csrf)])
def ingest_detection_event(organization_id: str, payload: DetectionEventIn, request: Request) -> dict:
    validate_origin_or_referer(request)
    if payload.organization_id != organization_id:
        raise HTTPException(status_code=400, detail="organization_id mismatch")
    incident = _repository(request).create_from_detection(parse_detection_event(payload.model_dump()))
    return _status_response(incident)


@router.get("/organizations/{organization_id}/incidents", dependencies=[Depends(require_permission("incidents.read"))])
def list_incidents(organization_id: str, request: Request, limit: int = DEFAULT_LIMIT, offset: int = 0, status: str | None = None, site_id: str | None = None, camera_id: str | None = None, zone_id: str | None = None, severity: str | None = None, created_from: str | None = None, created_to: str | None = None) -> dict:
    repo = _repository(request)
    created_from_dt = datetime.fromisoformat(created_from.replace("Z", "+00:00")) if created_from else None
    created_to_dt = datetime.fromisoformat(created_to.replace("Z", "+00:00")) if created_to else None
    filters = {"status": status, "site_id": site_id, "camera_id": camera_id, "zone_id": zone_id, "severity": severity, "created_from": created_from_dt, "created_to": created_to_dt}
    if hasattr(repo, "list_filtered"):
        items = repo.list_filtered(organization_id, limit=limit, offset=offset, **filters)
        total = repo.count_filtered(organization_id, **filters) if hasattr(repo, "count_filtered") else len(repo.list_filtered(organization_id, **filters))
        page = [incident_to_dict(incident) for incident in items]
        return {"items": page, "page_info": {"limit": max(1, min(limit, MAX_LIMIT)), "offset": max(0, offset), "total": total, "has_next": max(0, offset) + max(1, min(limit, MAX_LIMIT)) < total}}
    items = repo.list_by_organization(organization_id)
    return _paginate([incident_to_dict(incident) for incident in items], limit=limit, offset=offset)


@router.get("/organizations/{organization_id}/incidents/{incident_id}", dependencies=[Depends(require_permission("incidents.read"))])
def get_incident(organization_id: str, incident_id: str, request: Request) -> dict:
    try:
        return incident_to_dict(_repository(request).get(organization_id, incident_id))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="incident not found") from exc


@router.post("/organizations/{organization_id}/incidents/{incident_id}:acknowledge", dependencies=[Depends(require_permission("incidents.acknowledge")), Depends(require_csrf)])
def acknowledge_incident(organization_id: str, incident_id: str, request: Request, current_user=Depends(get_current_user)) -> dict:
    validate_origin_or_referer(request)
    try:
        return incident_to_dict(_repository(request).transition(organization_id, incident_id, IncidentStatus.ACKNOWLEDGED, current_user.user.id))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="incident not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/organizations/{organization_id}/incidents/{incident_id}:resolve", dependencies=[Depends(require_permission("incidents.resolve")), Depends(require_csrf)])
def resolve_incident(organization_id: str, incident_id: str, payload: IncidentReasonIn, request: Request, current_user=Depends(get_current_user)) -> dict:
    validate_origin_or_referer(request)
    try:
        return incident_to_dict(_repository(request).transition(organization_id, incident_id, IncidentStatus.RESOLVED, current_user.user.id, reason=payload.reason))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="incident not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400 if "requires reason" in str(exc) else 409, detail=str(exc)) from exc


@router.post("/organizations/{organization_id}/incidents/{incident_id}:dismiss", dependencies=[Depends(require_permission("incidents.dismiss")), Depends(require_csrf)])
def dismiss_incident(organization_id: str, incident_id: str, payload: IncidentReasonIn, request: Request, current_user=Depends(get_current_user)) -> dict:
    validate_origin_or_referer(request)
    try:
        return incident_to_dict(_repository(request).transition(organization_id, incident_id, IncidentStatus.DISMISSED, current_user.user.id, reason=payload.reason))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="incident not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400 if "requires reason" in str(exc) else 409, detail=str(exc)) from exc


@router.get("/organizations/{organization_id}/incidents/{incident_id}/audit-log", dependencies=[Depends(require_permission("audit.read"))])
def incident_audit_log(organization_id: str, incident_id: str, request: Request, limit: int = DEFAULT_LIMIT, offset: int = 0, action: str | None = None) -> dict:
    repo = _repository(request)
    items = repo.audit_logs(organization_id, incident_id)
    if action:
        items = [entry for entry in items if entry.action == action]
    return _paginate([entry.__dict__ for entry in items], limit=limit, offset=offset)

File: apps/web/src/api/incidents.ts
import { apiFetch } from './client'
import type { Metadata, PageInfo } from './types'

export type IncidentStatus = 'open' | 'acknowledged' | 'resolved' | 'dismissed'

export interface Incident {
  id: string
  organization_id: string
  site_id: string | null
  detection_event_id: string
  camera_id: string
  zone_id: string
  worker_id: string | null
  event_type: string
  severity: string
  summary: string
  confidence: number | null
  metadata: Metadata
  status: IncidentStatus
  created_at: string
  updated_at: string
  acknowledged_at: string | null
  resolved_at: string | null
  dismissed_at: string | null
}

export interface AuditLogEntry {
  id: string
  organization_id: string
  incident_id: string
  action: string
  from_status: string | null
  to_status: string
  actor: string
  created_at: string
  metadata: Metadata
}

export interface IncidentListResponse {
  items: Incident[]
  page_info: PageInfo
}

export interface AuditLogResponse {
  items: AuditLogEntry[]
  page_info: PageInfo
}

export interface IncidentListParams {
  limit?: number
  offset?: number
  status?: IncidentStatus
  site_id?: string
  camera_id?: string
  zone_id?: string
  severity?: string
  created_from?: string
  created_to?: string
}

export function listIncidents(organizationId: string, params: IncidentListParams = {}) {
  const query = new URLSearchParams()
  query.set('limit', String(params.limit ?? 50))
  query.set('offset', String(params.offset ?? 0))
  if (params.status) query.set('status', params.status)
  if (params.site_id) query.set('site_id', params.site_id)
  if (params.camera_id) query.set('camera_id', params.camera_id)
  if (params.zone_id) query.set('zone_id', params.zone_id)
  if (params.severity) query.set('severity', params.severity)
  if (params.created_from) query.set('created_from', params.created_from)
  if (params.created_to) query.set('created_to', params.created_to)
  return apiFetch<IncidentListResponse>(`/organizations/${organizationId}/incidents?${query.toString()}`)
}

export function getIncident(organizationId: string, incidentId: string) {
  return apiFetch<Incident>(`/organizations/${organizationId}/incidents/${incidentId}`)
}

export function getIncidentAuditLog(organizationId: string, incidentId: string, params: { limit?: number; offset?: number } = {}) {
  const query = new URLSearchParams()
  query.set('limit', String(params.limit ?? 50))
  query.set('offset', String(params.offset ?? 0))
  return apiFetch<AuditLogResponse>(`/organizations/${organizationId}/incidents/${incidentId}/audit-log?${query.toString()}`)
}

export function acknowledgeIncident(organizationId: string, incidentId: string) {
  return apiFetch<Incident>(`/organizations/${organizationId}/incidents/${incidentId}:acknowledge`, { method: 'POST' })
}

export function resolveIncident(organizationId: string, incidentId: string) {
  return apiFetch<Incident>(`/organizations/${organizationId}/incidents/${incidentId}:resolve`, { method: 'POST', body: JSON.stringify({ reason: 'Resolvido pelo painel operacional.' }) })
}

export function dismissIncident(organizationId: string, incidentId: string) {
  return apiFetch<Incident>(`/organizations/${organizationId}/incidents/${incidentId}:dismiss`, { method: 'POST', body: JSON.stringify({ reason: 'Descartado pelo painel operacional.' }) })
}
